File: SolaceApp/Services/FirestoreMessagingService.swift
import Foundation
import FirebaseFirestore
import SolaceCore

/// Firebase-backed implementation of `MessagingServicing`. Conversation
/// documents store both participant IDs (for querying) and their display
/// names (denormalized at creation time) so the conversation list doesn't
/// need a separate lookup per row.
// Firestore's SDK type is documented as thread-safe but doesn't yet carry
// Sendable conformance itself, hence @unchecked here.
final class FirestoreMessagingService: MessagingServicing, @unchecked Sendable {
    private let firestore = Firestore.firestore()

    func fetchCounselors() async throws -> [User] {
        let snapshot = try await firestore.collection("users")
            .whereField("role", isEqualTo: Role.counselor.rawValue)
            .getDocuments()
        return snapshot.documents.compactMap(Self.user(from:))
    }

    func startConversation(
        studentID: String,
        studentName: String,
        counselorID: String,
        counselorName: String
    ) async throws -> Conversation {
        // Filtered by participantIDs (not studentID/counselorID) so the query
        // shape matches what the security rule can prove safe — a compound
        // equality filter on studentID/counselorID can't be verified against
        // an `auth.uid in participantIDs` rule and gets rejected outright.
        let existing = try await firestore.collection("conversations")
            .whereField("participantIDs", arrayContains: studentID)
            .getDocuments()

        if let document = existing.documents.first(where: { ($0.data()["counselorID"] as? String) == counselorID }),
           let conversation = Self.conversation(from: document) {
            return conversation
        }

        let reference = firestore.collection("conversations").document()
        try await reference.setData([
            "studentID": studentID,
            "studentName": studentName,
            "counselorID": counselorID,
            "counselorName": counselorName,
            "participantIDs": [studentID, counselorID],
            "createdAt": FieldValue.serverTimestamp()
        ])
        return Conversation(
            id: reference.documentID,
            studentID: studentID,
            studentName: studentName,
            counselorID: counselorID,
            counselorName: counselorName
        )
    }

    func observeConversations(for userID: String) -> AsyncStream<[Conversation]> {
        AsyncStream { continuation in
            nonisolated(unsafe) let listener = firestore.collection("conversations")
                .whereField("participantIDs", arrayContains: userID)
                .order(by: "createdAt", descending: true)
                .addSnapshotListener { snapshot, _ in
                    guard let snapshot else { return }
                    continuation.yield(snapshot.documents.compactMap(Self.conversation(from:)))
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func observeMessages(conversationID: String) -> AsyncStream<[Message]> {
        AsyncStream { continuation in
            nonisolated(unsafe) let listener = firestore.collection("conversations").document(conversationID)
                .collection("messages")
                .order(by: "sentAt")
                .addSnapshotListener { snapshot, _ in
                    guard let snapshot else { return }
                    let messages = snapshot.documents.compactMap { Self.message(from: $0, conversationID: conversationID) }
                    continuation.yield(messages)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func sendMessage(conversationID: String, senderID: String, text: String) async throws {
        let conversationRef = firestore.collection("conversations").document(conversationID)
        try await conversationRef.collection("messages").addDocument(data: [
            "senderID": senderID,
            "text": text,
            "sentAt": FieldValue.serverTimestamp()
        ])
        try await conversationRef.setData([
            "lastMessageText": text,
            "lastMessageAt": FieldValue.serverTimestamp()
        ], merge: true)
    }

    private static func user(from document: QueryDocumentSnapshot) -> User? {
        let data = document.data()
        guard let role = Role(rawValue: data["role"] as? String ?? "") else { return nil }
        return User(
            id: document.documentID,
            email: data["email"] as? String ?? "",
            displayName: data["displayName"] as? String ?? "",
            role: role,
            bio: data["bio"] as? String,
            major: data["major"] as? String,
            academicYear: (data["academicYear"] as? String).flatMap(AcademicYear.init(rawValue:)),
            age: data["age"] as? Int
        )
    }

    private static func conversation(from document: QueryDocumentSnapshot) -> Conversation? {
        let data = document.data()
        guard
            let studentID = data["studentID"] as? String,
            let counselorID = data["counselorID"] as? String
        else { return nil }
        return Conversation(
            id: document.documentID,
            studentID: studentID,
            studentName: data["studentName"] as? String ?? "Student",
            counselorID: counselorID,
            counselorName: data["counselorName"] as? String ?? "Counselor",
            lastMessageText: data["lastMessageText"] as? String,
            lastMessageAt: (data["lastMessageAt"] as? Timestamp)?.dateValue(),
            createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date()
        )
    }

    private static func message(from document: QueryDocumentSnapshot, conversationID: String) -> Message? {
        let data = document.data()
        guard
            let senderID = data["senderID"] as? String,
            let text = data["text"] as? String
        else { return nil }
        return Message(
            id: document.documentID,
            conversationID: conversationID,
            senderID: senderID,
            text: text,
            sentAt: (data["sentAt"] as? Timestamp)?.dateValue() ?? Date()
        )
    }
}

File: SolaceAndroid/app/src/main/java/com/jonathanalumasa/solace/service/firebase/FirestoreMessagingService.kt
package com.jonathanalumasa.solace.service.firebase

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.SetOptions
import com.jonathanalumasa.solace.model.AcademicYear
import com.jonathanalumasa.solace.model.Conversation
import com.jonathanalumasa.solace.model.Message
import com.jonathanalumasa.solace.model.Role
import com.jonathanalumasa.solace.model.User
import com.jonathanalumasa.solace.service.MessagingService
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.tasks.await
import java.util.Date

/**
 * Firebase-backed implementation of [MessagingService]. Conversation documents
 * store both participant IDs (for querying) and their display names
 * (denormalized at creation time), so the conversation list doesn't need a
 * lookup per row.
 */
class FirestoreMessagingService(
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()
) : MessagingService {

    override suspend fun fetchCounselors(): List<User> =
        firestore.collection("users")
            .whereEqualTo("role", Role.COUNSELOR.rawValue)
            .get()
            .await()
            .documents
            .mapNotNull(::user)

    override suspend fun startConversation(
        studentID: String,
        studentName: String,
        counselorID: String,
        counselorName: String
    ): Conversation {
        // Filtered by participantIDs (not studentID/counselorID) so the query
        // shape matches what the security rule can prove safe — a compound
        // equality filter on studentID/counselorID can't be verified against
        // an `auth.uid in participantIDs` rule and gets rejected outright.
        val existing = firestore.collection("conversations")
            .whereArrayContains("participantIDs", studentID)
            .get()
            .await()
            .documents
            .firstOrNull { it.getString("counselorID") == counselorID }

        existing?.let { document ->
            conversation(document)?.let { return it }
        }

        val reference = firestore.collection("conversations").document()
        reference.set(
            mapOf(
                "studentID" to studentID,
                "studentName" to studentName,
                "counselorID" to counselorID,
                "counselorName" to counselorName,
                "participantIDs" to listOf(studentID, counselorID),
                "createdAt" to FieldValue.serverTimestamp()
            )
        ).await()

        return Conversation(
            id = reference.id,
            studentID = studentID,
            studentName = studentName,
            counselorID = counselorID,
            counselorName = counselorName
        )
    }

    override fun observeConversations(userID: String): Flow<List<Conversation>> =
        firestore.collection("conversations")
            .whereArrayContains("participantIDs", userID)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .snapshotListFlow(::conversation)

    override fun observeMessages(conversationID: String): Flow<List<Message>> =
        firestore.collection("conversations").document(conversationID)
            .collection("messages")
            .orderBy("sentAt")
            .snapshotListFlow { message(it, conversationID) }

    override suspend fun sendMessage(conversationID: String, senderID: String, text: String) {
        val conversationRef = firestore.collection("conversations").document(conversationID)
        conversationRef.collection("messages").add(
            mapOf(
                "senderID" to senderID,
                "text" to text,
                "sentAt" to FieldValue.serverTimestamp()
            )
        ).await()
        conversationRef.set(
            mapOf(
                "lastMessageText" to text,
                "lastMessageAt" to FieldValue.serverTimestamp()
            ),
            SetOptions.merge()
        ).await()
    }

    private fun user(document: DocumentSnapshot): User? {
        val role = Role.fromRaw(document.getString("role")) ?: return null
        return User(
            id = document.id,
            email = document.getString("email") ?: "",
            displayName = document.getString("displayName") ?: "",
            role = role,
            bio = document.getString("bio"),
            major = document.getString("major"),
            academicYear = AcademicYear.fromRaw(document.getString("academicYear")),
            age = document.getLong("age")?.toInt()
        )
    }

    private fun conversation(document: DocumentSnapshot): Conversation? {
        val studentID = document.getString("studentID") ?: return null
        val counselorID = document.getString("counselorID") ?: return null
        return Conversation(
            id = document.id,
            studentID = studentID,
            studentName = document.getString("studentName") ?: "Student",
            counselorID = counselorID,
            counselorName = document.getString("counselorName") ?: "Counselor",
            lastMessageText = document.getString("lastMessageText"),
            lastMessageAt = document.getDate("lastMessageAt"),
            createdAt = document.getDate("createdAt") ?: Date()
        )
    }

    private fun message(document: DocumentSnapshot, conversationID: String): Message? {
        val senderID = document.getString("senderID") ?: return null
        val text = document.getString("text") ?: return null
        return Message(
            id = document.id,
            conversationID = conversationID,
            senderID = senderID,
            text = text,
            sentAt = document.getDate("sentAt") ?: Date()
        )
    }
}
